### ISSUE TITLE
HTTP1/HTTP2 Proxy errors with CONNECT

### ISSUE THREAD

**Issue by arbarlow** [grpc/grpc-go#1243]
I posted this over at the Linkerd repo, but thought it would be useful to bring it up here too and see if we can resolve it somehow.

**Desired behaviour:**
Using linkerd as a http2 proxy with a gRPC client and gRPC server works transparently to the client and server.

**Actual behaviour:**
gRPC client attempts to do a http 1.0 `CONNECT` which fails.

**Setup**
gRPC Go: 1.4.0
Using the gRPC go[ 'hello world' examples](https://github.com/grpc/grpc-go/tree/master/examples/helloworld) (modify the client to send requests to 'world' instead of 'localhost'), then running..

`http_proxy=localhost:4140 go run main.go`

This causes the client to hang, as it will issue a http `CONNECT` to connect to the proxy, but gets no reply.

Wireshark shows this like the following..

![screen shot 2017-05-12 at 15 28 48](https://cloud.githubusercontent.com/assets/106357/26002502/c1a296ea-3727-11e7-86bf-a327b9ff6448.png)

Commenting out this line in gRPC-Go works as it uses the proxy, but skips the `CONNECT`: https://github.com/grpc/grpc-go/blob/master/proxy.go#L141

My linkerd config (essentially, it created a simple http2 based router/proxy) is the following..

```yaml
admin:
  port: 9990

routers:
- protocol: h2
  label: proxy
  experimental: true
  identifier:
    kind: io.l5d.header.token
  dtab: /svc => /$/inet/192.168.0.10/50051
  servers:
  - ip: 0.0.0.0
    port: 4140
```

`192.168.0.10/50051` being where my Go server is located

In summary, it would be useful to resolve this by either changing the way proxies work/skipping CONNECT is some circumstances or something similar. A timeout for the CONNECT should also probably be added. I'm happy to code the solution but I feel a discussion is definitely needed first!

**Comment by menghanl** [comment]
> Commenting out this line in gRPC-Go works as it uses the proxy

What do you mean by this? How does the proxy work if you comment out the support for `CONNECT`?

The support for `CONNECT` is part of the default dialer.
If you don't set `http_proxy`, it will not try to do `CONNECT`. If you do want to skip `CONNECT`, another way is to implement your own custom dialer. But I didn't get it why you want to skip the `CONNECT` handshake.

Since `CONNECT` is just one step of dialing, the dialing timeout applies here. You could do a `DialContext`, and call it with `WithBlock` dial option.

**Comment by arbarlow** [comment]
@menghanl thanks for taking a look.

Essentially the problem is that CONNECT is trying to tunnel http2 (grpc) over a http1 CONNECT proxy which doesn't work. So using the `http_proxy` env var is sort of an undocumented behaviour?

What I mean by commenting out the line is that when I used the proxy but don't do a CONNECT it works as expected.

**Comment by menghanl** [comment]
If I understand correctly, the HTTP Proxy server forwards the **TCP** connection to the desired destination. So http2 should work as expected.

Also, I may be wrong here, but is there an RFC talking about support for CONNECT with http2?

**Comment by markdroth** [comment]
Note that the http_proxy environment variable is specifically intended to enable use of HTTP CONNECT proxies -- i.e., it's for TCP-level proxies, not HTTP/2-level proxies.

If you want to use an HTTP/2-level proxy, then you can probably just point the client directly at the proxy's address.  There is no need to use the http_proxy environment variable.

For details on gRPC's TCP-level proxy support, see https://github.com/grpc/proposal/blob/master/A1-http-connect-proxy-support.md.

**Comment by arbarlow** [comment]
OK, thanks for the clarification.

I've left it to the linkerd guys to close their ticket but it would be nice to have http2_proxy support as well perhaps. Happy to close this and start something like that or kill it if need be.

**Comment by menghanl** [comment]
We don't see the need for this right now. So I'm closing this issue.
Please leave a comment if you still think this is necessary.
